### DISCUSSION TITLE
Adding another disjunction option while keeping all the existing ones from another definition

### DISCUSSION THREAD

**Discussion by umaasik** [cue-lang/cue#2997] [Q&A]
I'm new to CUE and thus my terminology may be wrong, so apologies in advance!

What I'm trying to achieve is while using this version of the [GitHub Actions workflow schema](https://raw.githubusercontent.com/SchemaStore/schemastore/5ffe36662a8fcab3c32e8fbca39c5253809e6913/src/schemas/json/github-workflow.json), is to augment the default elements for `runs-on`, which in CUE looks like the following:

```cue
// The type of machine to run the job on. The machine can be
// either a GitHub-hosted runner, or a self-hosted runner.
"runs-on": "macos-10.15" | "macos-11" | "macos-12" | "macos-latest" | "self-hosted" | "ubuntu-18.04" | "ubuntu-20.04" | "ubuntu-22.04" | "ubuntu-latest" | "windows-2019" | "windows-2022" | "windows-latest" | (["self-hosted", ...string] & [_, ...] | ["self-hosted", #machine, ...string] & [_, _, ...] | ["self-hosted", #architecture, ...string] & [_, _, ...] | ["self-hosted", #machine, #architecture, ...string] & [_, _, _, ...] | ["self-hosted", #architecture, #machine, ...string] & [_, _, _, ...]) & [...] | #stringContainingExpressionSyntax
```

The snippet above is from the CUE version of the schema (`cue import -f -p github -l #Workflow: jsonschema: - -o -`) in the file `github-workflow.cue` in the following directory structure:

```console
$ tree
.
├── cue.mod
│   ├── module.cue
│   ├── pkg
│   │   └── json.schemastore.org
│   │       └── github
│   │           └── github-workflow.cue
│   └── usr
│       └── json.schemastore.org
│           └── github
│               └── workflow.cue
└── workflows.cue
```

Under `workflows.cue` (plural, top-level) I'm able to successfully do the following:

```cue
package workflows

import "json.schemastore.org/github"

#javaWorkflow: github.#Workflow & {
	jobs: [string]: "runs-on": "ubuntu-latest"
}
```

But what I'd want is to input an arbitrary string to be able to reference self-hosted runners. Looking at the schema I noticed that it's possible to just write `["self-hosted", "my-arbitrary-string"]` and that would be fine, but when the workflow starts it gets stuck because there is no runner registered under "self-hosted". I could just register a new runner under that name, but I want to get better at CUE.

I'm looking for a way define an override for `runs-on` such that everything that is already defined under the schema will remain so, but that I am able to say that just "my-arbitrary-string" would also be okay to input. My `workflow.cue` (singular, nested under `usr`) looks like this:

```cue
package github

// TODO: drop when cuelang.org/issue/390 is fixed.
// Declare definitions for sub-schemas
#job:  ((#Workflow & {}).jobs & {x: _}).x
#step: ((#job & {steps:             _}).steps & [_])[0]

#NewNormalJob: #Workflow | *{
    jobs: #Workflow.#normalJob.jobs & {
        "runs-on": _ | "my-arbitrary-string"
    }
}
```

Ideally I'd love it if I didn't have to create a whole new definition (`#NewNormalJob`) and could just leverage the existing `#Workflow` definition, but that may be a pipe dream.

**Comment by jpluscplusm** [comment]
Just to check: you've read https://docs.github.com/en/actions/using-jobs/choosing-the-runner-for-a-job#choosing-self-hosted-runners, where it says:

> Although the self-hosted label is not required, we strongly recommend specifying it when using self-hosted runners to ensure that your job does not unintentionally specify any current or future GitHub-hosted runners

... and your requirement is to *allow* the consumer to do what they don't recommend: to *only* specify a single label, without also specifying the `self-hosted` label? (It'll either be doable in CUE or not - I'm just checking I've correctly understood the requirement!)

**Reply by umaasik** [reply]
You'd be right in the usual scenario of self-hosting runners, but we are using [GitHub ARC](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners-with-actions-runner-controller/) (sorry for not being more specific from the beginning) and for whatever reason it [does not support the creation of multiple labels](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners-with-actions-runner-controller/about-actions-runner-controller#using-arc-runners-in-a-workflow) and does not even set the `self-hosted` label on any registered runners; the runner is just stuck with a singular label with whatever the value to `runnerScaleSetName` is.

Currently we've worked around this by setting the value to just `${{ vars.SOME_VAR }}`, which satisfies the regular expression (see below), but I'm trying to find out if there's something I can do in CUE as well to make things a bit nicer, subjectively.

The regular expression:

```cue
#stringContainingExpressionSyntax: =~"""
	^.*\\$\\{\\{(.|[\r
	])*\\}\\}.*$
	"""
```
